lib/fuse_i.h

```diff
@@ -257,6 +257,15 @@ static inline int convert_to_conn_want_ext(struct fuse_conn_info *conn,
 					   uint64_t want_ext_default,
 					   uint32_t want_default)
 {
+	/* Check if want has bits set that want_ext doesn't.
+	 * In this case assume the application has used a mix of functions
+	 * and direct modification of want and copy those bits over to want_ext
+	 */
+	uint32_t want_diff = conn->want ^ fuse_lower_32_bits(conn->want_ext);
+
+	if ((conn->want_ext & want_diff) == 0)
+		conn->want_ext |= want_diff;
```

Comment:
Issue is that the application might delete default flags from either of both and this might add them back.

Reply:
Yes, that's unfortunate. It seems the whole `want_ext` stuff is too complex for a minor release as it breaks both ABI and API ways impossible to fix.

I wonder if this should be guarded by a feature flag (`#define FUSE_ONLY_WANT_EXT` or somesuch) so applications can declare that they don't touch `want`.

Reply:
Well, my personal wish was to entirely break the ABI and to make 'want' 64-bit. We had the .so bump anyway. Issue is that the creator of this project needs compatibility for an enterprise distro. 
I *think* your change in `fuse_apply_conn_info_opts` should be enough. I'm going to create test for these two functions (fuse_parse_conn_info_opts/fuse_apply_conn_info_opts).

If nothing works, we would need another .so bump (to .so.5) and do what I initially planned.

Reply:
No, it's not enough for the `ceph-fuse` case. That was my first attempt.

The sequence in that code is:

1. `fuse_parse_cmdline`
2. `fuse_parse_conn_info_opts`
3. In `fuse_lowlevel_ops->init` (`do_init`): (each guarded by a check against `capable`)
   1. `want |= FUSE_CAP_SPLICE_MOVE`
   2. `want |= FUSE_CAP_DONT_MASK`
   3. `want |= FUSE_CAP_EXPORT_SUPPORT`

Now `want` has up to three bits set which `want_ext` doesn't.

Reply:
I have to admit I'm a bit confused why my patch to `fuse_parse_conn_info_opts` is necessary in the first place as `conn.want` gets initialized from `conn.want_ext` in `_do_init`...

Reply:
Another finding: The docs for `fuse_apply_conn_info_opts` only mention `want`, not `want_ext`. So the current state of the code does not match the expectations of library users.

Reply:
Ok yeah, with fuse_parse_conn_info_opts() want and want_ext are changed from default ans then also other flags added to either of want or want_ext. So just updating there won't help. I think a solution is to move convert_to_conn_want_ext() into a C file and then to make want_ext_default and want_default global variables (or with setters) - fuse_parse_conn_info_opts() needs to update them. All not beautiful.

Reply:
See https://github.com/sur5r/libfuse/commit/7c84c67c696e222441581101c568b14cf42effde for a variant that works with `ceph-fuse` as well.

Reply:
That commit has the issue that it might add flags that the application might not want, not an ideal solution either. https://github.com/sur5r/libfuse/commit/7c84c67c696e222441581101c568b14cf42effde

And yeah, changing want works, because ceph sets want in its init method. But for example gvfsd already got update and uses fuse_set_feature_flag(), which sets conn_ext. Too late for the fuse API, but not ideal that structs are exposed at all - we need a time machine ;)

Reply:
@sur5r didn't manage to work on it yet - will do during a long flight tomorrow.

Reply:
> @sur5r didn't manage to work on it yet - will do during a long flight tomorrow.

Any news? It was two weeks before.

Reply:
https://github.com/libfuse/libfuse/pull/1223

Reply:
Sorry, hadn't managed to complete it during the flights and had been too busy the other days.